### ISSUE TITLE
[Question/Help required] Session Cookie - Server Side Session expiration config

### ISSUE THREAD

**Issue by Lukino** [DuendeSoftware/Support#588]
Using Duende IdentityServer 6.1.6 with .NET 6

My question are about SSO sessions.
I have configured Identity Server with ASP.NET Identity + EF without Server-Side-Sessions.

After successful login, in chrome I have two cookies:
  1. **.AspNetCore.Identity.Application** with expiration (14 days)
  2. **idsrv.session** with no expiration (Session)

When I try to login again less then 1 hour IS returns access token without login page and that's great.
But, if I try to login after 1 hour, it shows login page regarding of **.AspNetCore.Identity.Application** cookie expiration. And I assume that is somehow related **idsrv.session**, but I can't find exact logic how it works.

How can I extend this time? more preferable if I can configure it per client.

Now second question, I enabled Server Side Sessions.
I have same cookies and expirations in chrome and record in DB for server side session with expiration of 14 days (same as **.AspNetCore.Identity.Application** cookie ).
But when I try to login after 1 hour, IS redirects me on login screen.
Here is a details from Server Side Session record:
  Created: 2023-03-24 09:16:33.000
  Expires: 2023-04-07 09:16:33.000

Is this bug, that Server Side Session record has expiration after 14 days, but it gets expired after 1 hour?
How can I configure server side session lifetime ? more preferable if I can configure it per client.

To sum up, I want to extend SSO session. More preferable if I could configure lifetime per client.

**Comment by Lukino** [comment]
Identity Server configuration here:

            services
                .AddIdentity<User, IdentityRole<long>>(config =>
                    {
                        config.Password.RequireDigit = false;
                        config.Password.RequireLowercase = false;
                        config.Password.RequireNonAlphanumeric = false;
                        config.Password.RequireUppercase = false;
                        config.Password.RequiredLength = 6;
                    })
                .AddEntityFrameworkStores<DatabaseContext>()
                .AddUserStore<CustomUserStore>()
                .AddClaimsPrincipalFactory<UserClaimsPrincipalFactory>()
                .AddDefaultTokenProviders();

            services.AddIdentityServer(options =>
                {
                    options.ServerSideSessions.UserDisplayNameClaimType = "name";
                    options.Authentication.CookieLifetime= TimeSpan.FromMinutes(3);
                    options.ValidateTenantOnAuthorization = true;
                    options.KeyManagement.Enabled = false;

                    if (!string.IsNullOrEmpty(licenseKey))
                    {
                        options.LicenseKey = licenseKey;
                        logger.Information("License was found and configured for IS");
                    }
                    else
                    {
                        logger.Information("No license found");
                    }
                })
                .AddServerSideSessions()
                .AddConfigurationStoreCache()
                .AddConfigurationStore<DatabaseContext>(options => options.ResolveDbContextOptions = DatabaseContext.InitializeDbContextBuilderOptions)
                .AddOperationalStore<DatabaseContext>(options => options.ResolveDbContextOptions = DatabaseContext.InitializeDbContextBuilderOptions)
                .AddAspNetIdentity<User>()
                .AddProfileService<IdentityProfileService>()
                .AddResourceOwnerValidator<ResourceOwnerPasswordValidator<User>>()
                .AddCustomTokenRequestValidator<CustomTokenRequestValidator>()
                .AddSigningCredential(signingCredentialSettings.CertificateName);


            app.UseIdentityServer();
            app.UseAuthentication();
            app.UseAuthorization();

**Comment by StuFrankish** [comment]
You can configure token lifetimes per client;
https://docs.duendesoftware.com/identityserver/v6/reference/models/client

Have a look at the [#Token](https://docs.duendesoftware.com/identityserver/v6/reference/models/client/#token) and [#Refresh Token](https://docs.duendesoftware.com/identityserver/v6/reference/models/client/#refresh-token) sections for lifetime settings.

Here's a snippet of `Duende.IdentityServer.Models.Client`
```c#
/// <summary>
/// Lifetime of identity token in seconds (defaults to 300 seconds / 5 minutes)
/// </summary>
public int IdentityTokenLifetime { get; set; } = 300;

/// <summary>
/// Signing algorithm for identity token. If empty, will use the server default signing algorithm.
/// </summary>
public ICollection<string> AllowedIdentityTokenSigningAlgorithms { get; set; } = new HashSet<string>();

/// <summary>
/// Lifetime of access token in seconds (defaults to 3600 seconds / 1 hour)
/// </summary>
public int AccessTokenLifetime { get; set; } = 3600;

/// <summary>
/// Lifetime of authorization code in seconds (defaults to 300 seconds / 5 minutes)
/// </summary>
public int AuthorizationCodeLifetime { get; set; } = 300;

/// <summary>
/// Maximum lifetime of a refresh token in seconds. Defaults to 2592000 seconds / 30 days
/// </summary>
public int AbsoluteRefreshTokenLifetime { get; set; } = 2592000;

/// <summary>
/// Sliding lifetime of a refresh token in seconds. Defaults to 1296000 seconds / 15 days
/// </summary>
public int SlidingRefreshTokenLifetime { get; set; } = 1296000;
```

**Comment by josephdecock** [comment]
IdentityServer can either use a cookie authentication scheme that you set up yourself, or if you don't, it will create one for you.

`options.Authentication.CookieLifetime= TimeSpan.FromMinutes(3);` only applies if you're not creating a cookie scheme of your own, so to change the lifetime of your sessions, you need to set the `ExpireTimeSpan` in the cookie handler that you're creating.

**Comment by StuFrankish** [comment]
> `options.Authentication.CookieLifetime= TimeSpan.FromMinutes(3);` only applies if you're not creating a cookie scheme of your own, so to change the lifetime of your sessions, you need to set the `ExpireTimeSpan` in the cookie handler that you're creating.

Is there reference documentation for that @josephdecock ?
I can't see my own implementation of IdentityServer using `ExpireTimeSpan` so is mine only using the client specific configuration?

**Comment by josephdecock** [comment]
The CookieLifetime option is documented in the reference section on options [here](here: https://docs.duendesoftware.com/identityserver/v6/reference/options/#authentication), and demonstrated in the [cookie handler section on sessions](https://docs.duendesoftware.com/identityserver/v6/ui/login/session/#cookie-handler-configuration), but I'm always interested to hear when folks have trouble finding things in the docs. Any suggestions on this topic are welcome.

CookieLifetime is separate from the per-client configuration of token lifetimes. It controls the lifetime of the IdentityServer session - how long will you have single sign on across clients. That session and its duration and other settings aren't dependent on which clients happen to be interacting with IdentityServer. 

If what you're trying to accomplish is that some clients require a more recent authentication time, then you can use [step up](https://datatracker.ietf.org/doc/draft-ietf-oauth-step-up-authn-challenge/) to accomplish that. IdentityServer supports step up, and we have a sample of how to implement it [here](https://docs.duendesoftware.com/identityserver/v6/samples/ui/#step-up).
